concepts/functions/about.md

```diff
@@ -211,51 +211,128 @@ Now, we can call the functions in the order we want.
+If a variable (_or name_) is defined inside a function, it will be only accessible _inside_ the function scope (_or local namespace_), even if there is a non-local variable with the same name.
+Variables defined outside a function at the module level are considered in the _global namespace_.
+Variables defined outside a function but _inside_ an enclosing function or class are in the _nonlocal namespace_.
 
-```python
->>> num = 30
->>> def random_function():
-        num = 10
-        print('Inside function:', num)
-```
+Python uses the [LEGB Rule][LEGB Rule] (**L**ocal, **E**nclosing, **G**lobal, **B**uilt-in) to resolve variable names when a program is running:
+
+1. Lookup in the **local** (or _function_) namespace.
+2. Lookup in the **enclosing** (or _nonlocal_) namespace if a name is not found in local.
+3. Lookup in the **global** namespace (_module level_) if the name is not found in nonlocal/enclosing.
+4. Lookup in the **built-in** (_program or python-wide_) namespace if the name is not found in global.
 
-As `num` is defined inside the `random_function`, it is limited to the scope of the `random_function` only.
-Calling the function will not alter the value of the variable outside the function.
+If the name remains unresolved, Python will raise a `NameError exception`.
 
 ```python
->>> num = 30
+# Global namespace.
+general_favorite = "apples"
+
+
+def alices_favorite():
+   # Local namespace.  
+   favorite = "cherries"
+   
+   # This works because Python will eventually find general_favorite in the global namespace.
+   # Python will find 'print' in the built-in namespace.
+   print(f'Alice has always liked {favorite}, but most people like {general_favorite}.')
+
+   
+def our_favorite_fruits():
+    # Enclosing or nonlocal namespace.
+    yours = "peaches"
+    
+    def my_favorite():
+        # Local namespace.
+        mine = "pears"
+        
+        # This works because Python will eventually find 'yours' in the enclosing/nonlocal namespace.
+        print(f'My favorite is {mine}, but you like {yours} instead.')
+        
+        # This works because Python will eventually find 'general_favorite' in the global namespace.
+        print(f'Everyone seems to like {general_favorite}')
+    
+    # This function is in the local namespace of the 'our_favorite_fruits' function.  
+    my_favorite()
+
+    
+# This will raise NameError: name 'favorite' is not defined, because the variable favorite is local to alices_favorite.
+print(favorite)
 
-# regardless of whether we call the function or not, the value of num will be 30
->>> random_function()
-Inside function: 10
->>> num
-30
 ```
 
-We can access the variable inside the outer function using the `global` keyword.
+If we want to make a variable name accessible _outside_ the local function scope (_or modify a variable defined outside the function scope_), we need to use either the [`global`][global] or [`nonlocal`][nonlocal] keywords.
```

Comment:
Overall, this was great work! I especially like the `LEGB`.

Reply:
I think the other thing that we may need to separate out is that this goes both directions.  You can pull in and modify a variable from an enclosing scope...but you can also _hoist_ a local variable to the global scope, which I don't think we have a code example for.

Reply:
Yes, it might be helpful to comment it such, since what it references is far up the page from where it errors. Something like

```python
        # Attempting to reference im_global will error because it is already
        # referenced in outer_function
        global im_global
```
Also, if `im_global` is just _read_ in `outer_function` it will error when trying to then reference `im_global` as global

```python
    print(im_global)
    # Attempting to reference the global variable via global keyword
    # will error since `im_global` has already been used
    global im_global
```

Reply:
Well .. not _really_ a hoist, but you can make a `global` variable within a function by using `global`.  You can also make a variable within a nested function nonlocal by using `nonlocal`.

Reply:
> _Yes, it might be helpful to comment it such, since what it references is far up the page from where it errors. Something like_

Yup.  It's confusing (_even after I've read a bunch of examples_).  I agree:  let's break this up so that its clearer.  I'll try to do that later today, after I get some more done on the roadmap doc.  

I am also happy to let you edit it (_I think the [GitHub desktop client](https://desktop.github.com/) will let you make changes/commit directly to this branch_), if you'd rather not go back and forth here in the web UI.

Reply:
Not to be outrageous, but it might be worth considering if `global` and `nonlocal` belong to the `function` concept, or to a separate `scope` concept. I mean, if there is a `function` concept exercise, it's not likely to involve `global` or `nonlocal`. Giving them that info in the `function` concept is kind of like giving a child wooden blocks along with a couple of matchbooks.

Reply:
Oh, I see what you mean about declaring `global` or `nonlocal` in an enclosed scope making it accessible in the enclosing scope.

Reply:
> _Not to be outrageous, but it might be worth considering if `global` and `nonlocal` belong to the `function` concept, or to a separate `scope` concept. I mean, if there is a `function` concept exercise, it's not likely to involve `global` or `nonlocal`._

I don't think that's an outrageous idea - it is a fully legitimate question we should consider.  Some of the things I'm thinking as I consider that:
* If it was `scope` in all its forms, we'd need to put it in the tree somewhere after `comprehensions` (_they create their own scope_)`classes`, `functions`, `lambdas` (_again, they create their own scope_), `with` (context handlers), and `modules`.  Or, at the very least `classes`, `functions` and `lambdas`.
* What kind of concept exercise would we be able to create that touched on all that scoping?  What kinds of practice exercises?  Feels like `scoping` is way too broad a concept to really do a good job, but I am probably just not seeing it right now.
* If we remove talk of scope from `functions`, and we remove `arguments` from `functions`, and we remove explicit `None` from `functions` -- what are we left with?  

   The `basics` concept already covers base function definition, `return` and implicitly returning `None`.  We have follow-on concepts that cover `function-arguments`, `lambdas`, `higher-order functions`, `None` and `functools`. 

   That leaves us with reiteration (_not a bad thing_) of `function` definition, `functions as callable objects`, and `function special attributes`.  _**Are those three topics enough for a functions concept exercise?**_  And what practice exercises would we need for it?

<br>

> _Giving them that info in the `function` concept is kind of like giving a child wooden blocks along with a couple of matchbooks._

...aaaand you've just summed up many of the corners of Python. 😆   But on a more serious note - this would be the time to warn about fire-setting.  `Functions` was not a concept that was intended to appear right after `basics`. 

 Instead, the vision was to have students get all the "core" Python stuff practiced **first**.  So prerequisites to the `functions` concept exercise would include everything before `loops`, `loops` and probably all the other data structures that come before `classes`.  That's about 8-10 concept exercises, including `basics` - probably more, since we will be inserting `sequences`, `list-comprehensions`, and `other-comprehensions` into the tree as well.

Presumably, after all those concept exercises (_and maybe additional practice exercises_), the students would have quite a bit of basic Python under their belt before hitting this `functions` concept, so they'd then benefit from some instruction on some of the dangerous but (perhaps) useful tools for `functions` they have at hand.

Reply:
> _Oh, I see what you mean about declaring `global` or `nonlocal` in an enclosed scope making it accessible in the enclosing scope._

Yup.  It is indeed weird (_and a foot-gun if not used very carefully_).

Reply:
> If we remove talk of scope from functions, and we remove arguments from functions, and we remove explicit None from functions -- what are we left with?

Well, I'm not advocating removing arguments or the implicit `None` return, as those could conceivably be part of a concept exercise. At a minimum, they are things we want the student to know and make use of. We don't necessarily want the students to make use of `global` and `nonlocal`, so maybe best to keep shtum about it?

Reply:
Maybe scope could be gotten into in terms of encapsulation and variable hiding/shadowing. `global` and `nonlocal` could be mentioned with links to their docs as ways to modify a variable not within its scope, but say that use of them is discouraged because it breaks encapsulation. Examples could be given for encapsulation and variable shadowing, but maybe no need to example how to do anything it's not recommended to do (?)

Reply:
I think that sounds reasonable.  How would you like to proceed here?  I can pull out the `nonlocal` and `global` code examples and leave in the LEGB rule/scoping.  Then we can then merge this PR, and then work on the "warnings" around `global`/`nonlocal`, add any stuff around shadowing/hiding and additional (if needed) examples on encapsulation. What do you think?

Reply:
Sure, that sounds okay.